Unable to get provider leakcanary.internal.AppWatcherInstaller$MainProcess:

### Description


Failure on building debug apk. Unable to get provider leakCanary when it tries to install  AppWatcher


### Version Information

* LeakCanary version: 2.0
* Android OS version: 6.0.1
* Gradle version: 4.1

### Additional Information

sample project that reproduces issue: https://drive.google.com/file/d/1DSTrQOqsiYNZ71zJUK_DjbihFCxNSR6Z/view?usp=sharing

Logcat:

```
12-06 15:42:23.987 15540-15540/? I/art: Late-enabling -Xcheck:jni
    Late-enabling JIT
12-06 15:42:24.009 15540-15540/? I/art: JIT created with code_cache_capacity=2MB compile_threshold=1000
12-06 15:42:24.070 15540-15540/com.app.tablesign W/ActivityThread: Application com.app.tablesign is waiting for the debugger on port 8100...
12-06 15:42:24.074 15540-15540/com.app.tablesign I/System.out: Sending WAIT chunk
12-06 15:42:24.245 15540-15550/com.app.tablesign I/art: Debugger is active
12-06 15:42:24.275 15540-15540/com.app.tablesign I/System.out: Debugger has connected
    waiting for debugger to settle...
12-06 15:42:24.475 15540-15540/com.app.tablesign I/System.out: waiting for debugger to settle...
12-06 15:42:24.675 15540-15540/com.app.tablesign I/System.out: waiting for debugger to settle...
12-06 15:42:24.876 15540-15540/com.app.tablesign I/System.out: waiting for debugger to settle...
12-06 15:42:25.076 15540-15540/com.app.tablesign I/System.out: waiting for debugger to settle...
12-06 15:42:25.277 15540-15540/com.app.tablesign I/System.out: waiting for debugger to settle...
12-06 15:42:25.477 15540-15540/com.app.tablesign I/System.out: waiting for debugger to settle...
12-06 15:42:25.677 15540-15540/com.app.tablesign I/System.out: waiting for debugger to settle...
12-06 15:42:25.878 15540-15540/com.app.tablesign I/System.out: debugger has settled (1497)
12-06 15:42:25.899 15540-15540/com.app.tablesign W/System: ClassLoader referenced unknown path: /data/app/com.app.tablesign-1/lib/arm
12-06 15:42:25.972 15540-15540/com.app.tablesign D/LeakCanary: Installing AppWatcher
12-06 15:42:25.984 15540-15540/com.app.tablesign I/art: Rejecting re-init on previously-failed class java.lang.Class<leakcanary.internal.AndroidSupportFragmentDestroyWatcher$fragmentLifecycleCallbacks$1>
12-06 15:42:25.985 15540-15540/com.app.tablesign I/art: Rejecting re-init on previously-failed class java.lang.Class<leakcanary.internal.AndroidSupportFragmentDestroyWatcher$fragmentLifecycleCallbacks$1>
12-06 15:42:25.989 15540-15540/com.app.tablesign I/art: Rejecting re-init on previously-failed class java.lang.Class<leakcanary.internal.AndroidSupportFragmentDestroyWatcher$fragmentLifecycleCallbacks$1>
    Rejecting re-init on previously-failed class java.lang.Class<leakcanary.internal.AndroidSupportFragmentDestroyWatcher$fragmentLifecycleCallbacks$1>
12-06 15:42:25.991 15540-15540/com.app.tablesign I/art: Rejecting re-init on previously-failed class java.lang.Class<leakcanary.internal.AndroidSupportFragmentDestroyWatcher$fragmentLifecycleCallbacks$1>
12-06 15:42:25.991 15540-15540/com.app.tablesign D/AndroidRuntime: Shutting down VM
    
    
    --------- beginning of crash
12-06 15:42:25.993 15540-15540/com.app.tablesign E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.app.tablesign, PID: 15540
    java.lang.RuntimeException: Unable to get provider leakcanary.internal.AppWatcherInstaller$MainProcess: java.lang.reflect.InvocationTargetException
        at android.app.ActivityThread.installProvider(ActivityThread.java:5156)
        at android.app.ActivityThread.installContentProviders(ActivityThread.java:4748)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4688)
        at android.app.ActivityThread.-wrap1(ActivityThread.java)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1405)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:148)
        at android.app.ActivityThread.main(ActivityThread.java:5417)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:742)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:632)
     Caused by: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Constructor.newInstance(Native Method)
        at leakcanary.internal.FragmentDestroyWatcher.getWatcherIfAvailable(FragmentDestroyWatcher.kt:99)
        at leakcanary.internal.FragmentDestroyWatcher.install(FragmentDestroyWatcher.kt:62)
        at leakcanary.internal.InternalAppWatcher.install(InternalAppWatcher.kt:70)
        at leakcanary.internal.AppWatcherInstaller.onCreate(AppWatcherInstaller.kt:35)
        at android.content.ContentProvider.attachInfo(ContentProvider.java:1748)
        at android.content.ContentProvider.attachInfo(ContentProvider.java:1723)
        at android.app.ActivityThread.installProvider(ActivityThread.java:5153)
        at android.app.ActivityThread.installContentProviders(ActivityThread.java:4748) 
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4688) 
        at android.app.ActivityThread.-wrap1(ActivityThread.java) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1405) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:148) 
        at android.app.ActivityThread.main(ActivityThread.java:5417) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:742) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:632) 
     Caused by: java.lang.NoClassDefFoundError: leakcanary.internal.AndroidSupportFragmentDestroyWatcher$fragmentLifecycleCallbacks$1
        at leakcanary.internal.AndroidSupportFragmentDestroyWatcher.<init>(AndroidSupportFragmentDestroyWatcher.kt:30)
        at java.lang.reflect.Constructor.newInstance(Native Method) 
        at leakcanary.internal.FragmentDestroyWatcher.getWatcherIfAvailable(FragmentDestroyWatcher.kt:99) 
        at leakcanary.internal.FragmentDestroyWatcher.install(FragmentDestroyWatcher.kt:62) 
        at leakcanary.internal.InternalAppWatcher.install(InternalAppWatcher.kt:70) 
        at leakcanary.internal.AppWatcherInstaller.onCreate(AppWatcherInstaller.kt:35) 
        at android.content.ContentProvider.attachInfo(ContentProvider.java:1748) 
        at android.content.ContentProvider.attachInfo(ContentProvider.java:1723) 
        at android.app.ActivityThread.installProvider(ActivityThread.java:5153) 
        at android.app.ActivityThread.installContentProviders(ActivityThread.java:4748) 
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4688) 
        at android.app.ActivityThread.-wrap1(ActivityThread.java) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1405) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:148) 
        at android.app.ActivityThread.main(ActivityThread.java:5417) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:742) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:632) 

```

--- a/leakcanary-object-watcher-android/src/main/java/leakcanary/internal/FragmentDestroyWatcher.kt
+++ b/leakcanary-object-watcher-android/src/main/java/leakcanary/internal/FragmentDestroyWatcher.kt
@@ -92,10 +92,10 @@ internal object FragmentDestroyWatcher {
   ): ((Activity) -> Unit)? {
 
     return if (classAvailable(fragmentClassName) &&
-            classAvailable(watcherClassName)
+        classAvailable(watcherClassName)
     ) {
       val watcherConstructor = Class.forName(watcherClassName)
-              .getDeclaredConstructor(ObjectWatcher::class.java, Function0::class.java)
+          .getDeclaredConstructor(ObjectWatcher::class.java, Function0::class.java)
       @Suppress("UNCHECKED_CAST")
       watcherConstructor.newInstance(objectWatcher, configProvider) as (Activity) -> Unit
 
@@ -108,7 +108,15 @@ internal object FragmentDestroyWatcher {
     return try {
       Class.forName(className)
       true
-    } catch (e: ClassNotFoundException) {
+    } catch (e: Throwable) {
+      // e is typically expected to be a ClassNotFoundException
+      // Unfortunately, prior to version 25.0.2 of the support library the
+      // FragmentManager.FragmentLifecycleCallbacks class was a non static inner class.
+      // Our AndroidSupportFragmentDestroyWatcher class is compiled against the static version of
+      // the FragmentManager.FragmentLifecycleCallbacks class, leading to the
+      // AndroidSupportFragmentDestroyWatcher class being rejected and a NoClassDefFoundError being
+      // thrown here. So we're just covering our butts here and catching everything, and assuming
+      // any throwable means "can't use this". See https://github.com/square/leakcanary/issues/1662
       false
     }
   }

